compiler/rustc_infer/src/errors/mod.rs

```diff
@@ -497,3 +498,198 @@ pub struct MismatchedStaticLifetime<'a> {
     #[subdiagnostic]
     pub trait_subdiags: Vec<TraitSubdiag>,
 }
+
+#[derive(Diagnostic)]
+#[diag(infer::explicit_lifetime_required, code = "E0621")]
+pub struct ExplicitLifetimeRequired<'a> {
+    #[primary_span]
+    #[label]
+    pub span: Span,
+    pub ident_kind: &'static str,
+    pub simple_ident: String,
+    pub named: String,
+
+    #[suggestion(
+        infer::explicit_lifetime_required_sugg,
+        code = "{new_ty}",
+        applicability = "unspecified"
+    )]
+    pub new_ty_span: Span,
+    #[skip_arg]
+    pub new_ty: Ty<'a>,
+}
+
+#[derive(Subdiagnostic)]
+pub enum ActualImplExplNotes {
+    // Field names have to be different across all variants
+    #[note(infer::actual_impl_expl_1)]
+    NoteOne {
```

Comment:
It would be better if these had more descriptive names.

Reply:
Right, fixed this one.